pkg/lexer/lexer.go

```diff
@@ -1,5 +1,215 @@
+	return sb.String()
+}
+
+type Lexer struct {
+	input []rune
+	r     rune // current rune under examination
+	pos   int  // current position in input (points to current rune)
+	line  int
+	col   int
+}
+
+func New(input string) *Lexer {
+	return &Lexer{input: []rune(input), pos: -1, line: 1}
+}
+
+func (l *Lexer) readRune() {
+	l.pos += 1
+	l.col += 1
+	if l.r == '\n' {
+		l.line += 1
+		l.col = 1
+	}
+	if l.pos >= len(l.input) {
+		l.r = 0
+		return
+	}
+	l.r = l.input[l.pos]
+}
+
+func (l *Lexer) skipWhitespace() {
+	for l.r == ' ' || l.r == '\t' || l.r == '\r' {
+		l.readRune()
+	}
+}
+
+func (l *Lexer) peekRune() rune {
+	if l.pos+1 >= len(l.input) {
+		return 0
+	}
+	return l.input[l.pos+1]
+}
+
+func (l *Lexer) Next() *Token {
+	l.readRune()
+	l.skipWhitespace()
+
+	tok := &Token{
+		Offset: l.pos,
+		Line:   l.line,
+		Col:    l.col,
+	}
+	switch l.r {
+	case '=':
+		if l.peekRune() == '=' {
+			l.readRune()
+			return tok.SetType(EQ)
+		}
+		return tok.SetType(ASSIGN)
+	case '+':
+		return tok.SetType(PLUS)
+	case '-':
+		return tok.SetType(MINUS)
+	case '!':
+		if l.peekRune() == '=' {
+			l.readRune()
+			return tok.SetType(NOT_EQ)
+		}
+		return tok.SetType(BANG)
+	case '/':
+		if l.peekRune() == '/' {
+			return tok.SetType(COMMENT).SetLiteral(l.readComment())
+		}
+		return tok.SetType(SLASH)
+	case '*':
+		return tok.SetType(ASTERISK)
+	case '<':
+		if l.peekRune() == '=' {
+			l.readRune()
+			return tok.SetType(LTEQ)
+		}
+		return tok.SetType(LT)
+	case '>':
+		if l.peekRune() == '=' {
+			l.readRune()
+			return tok.SetType(GTEQ)
+		}
+		return tok.SetType(GT)
+	case ':':
+		if l.peekRune() == '=' {
+			l.readRune()
+			return tok.SetType(DECLARE)
+		}
+		return tok.SetType(COLON)
+	case '{':
+		return tok.SetType(LCURLY)
+	case '}':
+		return tok.SetType(RCURLY)
+	case '(':
+		return tok.SetType(LPAREN)
+	case ')':
+		return tok.SetType(RPAREN)
+	case '[':
+		return tok.SetType(LBRACKET)
+	case ']':
+		return tok.SetType(RBRACKET)
+	case '\n':
+		return tok.SetType(NL)
+	case '.':
+		if l.peekRune() == '.' && l.peekRune2() == '.' {
+			l.readRune()
+			l.readRune()
+			return tok.SetType(DOT3)
+		}
+		return tok.SetType(DOT)
+	case '"':
+		literal := l.readString()
+		// EOF or NL before closing `"`
+		if l.r != '"' {
+			return tok.SetType(ILLEGAL).SetLiteral(`"`)
+		}
+		return tok.SetType(STRING_LIT).SetLiteral(literal)
+	case 0:
+		return tok.SetType(EOF)
+	}
+	if isLetter(l.r) {
+		lit := l.readIdent()
+		tokenType := LookupKeyword(lit)
+		if tokenType == IDENT {
+			return tok.SetType(IDENT).SetLiteral(lit)
+		}
+		return tok.SetType(tokenType)
+	}
+	if isDigit(l.r) {
+		return tok.SetType(NUM_LIT).SetLiteral(l.readNum())
+	}
+
+	return tok.SetType(ILLEGAL).SetLiteral(string(l.r))
+}
+
+func (l *Lexer) peekRune2() rune {
+	if l.pos+2 >= len(l.input) {
+		return 0
+	}
+	return l.input[l.pos+2]
+}
+
+func (l *Lexer) readComment() string {
+	pos := l.pos
+	pr := l.peekRune()
+	for pr != 0 && pr != '\n' {
```

Comment:
You could write this loop as:
```
        for pr := l.peekRune(); pr != 0 && pr != '\n'; pr = l.peekRune() {
                l.readRune()
        }
```

Not sure if you feel that is any clearer. It's no fewer expressions, but two of them can go onto the `for` line.

Reply:
i tried this a couple of times with the readString function and gave up. looks goo here.
Done.